File: examples/governance/risk_assessment_domain/lib/risk_assessment_domain/assessment/assessment.rb
require 'hecks/model'

module RiskAssessmentDomain
  class Assessment
    autoload :Finding, "risk_assessment_domain/assessment/finding"
    autoload :Mitigation, "risk_assessment_domain/assessment/mitigation"
    autoload :Lifecycle, "risk_assessment_domain/assessment/lifecycle"

    include Hecks::Model

    attribute :model_id
    attribute :assessor_id
    attribute :risk_level
    attribute :bias_score
    attribute :safety_score
    attribute :transparency_score
    attribute :overall_score
    attribute :submitted_at
    attribute :findings, default: [], freeze: true
    attribute :mitigations, default: [], freeze: true
    attribute :status

    # State predicates — see lifecycle.rb for full state machine
    def pending?; status == "pending"; end
    def submitted?; status == "submitted"; end
    def rejected?; status == "rejected"; end

    VALID_RISK_LEVEL = ["low", "medium", "high", "critical"].freeze unless defined?(VALID_RISK_LEVEL)

    private

    def validate!
      raise ValidationError, "model_id can't be blank" if model_id.nil? || (model_id.respond_to?(:empty?) && model_id.empty?)
      raise ValidationError, "assessor_id can't be blank" if assessor_id.nil? || (assessor_id.respond_to?(:empty?) && assessor_id.empty?)
      if risk_level && !VALID_RISK_LEVEL.include?(risk_level)
        raise ValidationError, "risk_level must be one of: #{VALID_RISK_LEVEL.join(', ')}, got: #{risk_level}"
      end
    end

    def check_invariants!
      raise InvariantError, "scores must be between 0 and 1" unless instance_eval(&proc { [bias_score, safety_score, transparency_score, overall_score].all? { |s|
s.nil? || (s >= 0.0 && s <= 1.0)
} })
    end
  end
end

File: examples/governance_static_go/domain/assessment.go
package domain

import (
	"time"
	"github.com/google/uuid"
	"fmt"
)

type Assessment struct {
	ID string `json:"id"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	ModelId string `json:"model_id"`
	AssessorId string `json:"assessor_id"`
	RiskLevel string `json:"risk_level"`
	BiasScore float64 `json:"bias_score"`
	SafetyScore float64 `json:"safety_score"`
	TransparencyScore float64 `json:"transparency_score"`
	OverallScore float64 `json:"overall_score"`
	SubmittedAt time.Time `json:"submitted_at"`
	Findings []Finding `json:"findings"`
	Mitigations []Mitigation `json:"mitigations"`
	Status string `json:"status"`
}

func NewAssessment(modelId string, assessorId string, riskLevel string, biasScore float64, safetyScore float64, transparencyScore float64, overallScore float64, submittedAt time.Time, findings []Finding, mitigations []Mitigation, status string) *Assessment {
	a := &Assessment{
		ID:        uuid.New().String(),
		ModelId: modelId,
		AssessorId: assessorId,
		RiskLevel: riskLevel,
		BiasScore: biasScore,
		SafetyScore: safetyScore,
		TransparencyScore: transparencyScore,
		OverallScore: overallScore,
		SubmittedAt: submittedAt,
		Findings: findings,
		Mitigations: mitigations,
		Status: status,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}
	return a
}

func (a *Assessment) Validate() error {
	if a.ModelId == "" {
		return &ValidationError{Field: "model_id", Message: "model_id can't be blank"}
	}
	if a.AssessorId == "" {
		return &ValidationError{Field: "assessor_id", Message: "assessor_id can't be blank"}
	}
	if a.RiskLevel != "" {
		validRiskLevel := map[string]bool{"low": true, "medium": true, "high": true, "critical": true}
		if !validRiskLevel[a.RiskLevel] {
			return &ValidationError{Field: "risk_level", Message: fmt.Sprintf("risk_level must be one of: low, medium, high, critical, got: %s", a.RiskLevel)}
		}
	}
	// invariant: scores must be between 0 and 1
	return nil
}
